### PR TITLE
Added several new functionalities and some minor fixes

### PR DESCRIPTION
**Pull request by edelkas**
Hey this is Eddy. Apart from some minor changes, here's the main contributions of this request:

* Addition:  Maxed levels. Returns a list of all levels with at least 20 ties for 0th.
* Addition: Maxable levels. Returns a list of levels with the most ties for 0th (unless its maxed) in descending order.
* Addition: Community scores. Returns the community's total level / episode scores, difference, averages, and some more stuff.
* Addition: Cleanest and dirtiest episodes, that is, with less / more difference between the ep 0th and the 5 level 0ths. This is related to issue #46 
* Addition: Episode ownages.
* Addition: Tied 0th rankings and average rank rankings.
* Enhancement: Customizable score list. You can now ask for a list of the top n scores, or the bottom n scores, or 0th scores, or any combination thereof.

You may also notice a couple of fixes:

* When migrating the database the method "change_tables" in file 15 is overriding the Rails method, and this causes problems later on, e.g., with file 16 in the migration, so I changed the name.
* When seeding the database I got rid of ultimate, since that's being done in the migration and will cause a primary key constraint violation error. As a suggestion, it could be considered to unify and put everything either in the migration or in the seeding.
* I've (hopefully) taken care of issue #41 regarding inne's commands.

I'm cooking more interesting ideas, I'll keep you posted.

### CODE REVIEW

**Comment by liam-mitchell** [comment]
I like it. Haven't had a chance to review all the code yet, so I'll surely have some comments on the implementation, but thank you for doing all this - all of these seem like really good additions, and it's nice to have someone working through the setup steps and shaking out some issues in there. I'll probably message you on discord at some point to pick your brains about problems you had setting up and to talk about mysql setup (since I ended up changing to mysql over sqlite, due to sqlite's issues with multithreaded writes crashing sometimes when updating) - but these commits look great and I'm excited to take a look :)

File: db/migrate/15_change_tab_to_enum.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@ def tab(prefix)
   }[prefix]
 end
 
-def change_table(type, table)
+def modify_table(type, table)
```
Comment (liam-mitchell) [inline db/migrate/15_change_tab_to_enum.rb:12]: whooops... 😊

File: messages.rb [hunk 2]
```diff
@@ -64,6 +64,11 @@ def parse_rank(msg)
   rank ? rank.to_i : nil
 end
 
+def parse_bottom_rank(msg)
+  rank = msg[/bottom\s*([0-9][0-9]?)/i, 1]
+  rank ? 20-rank.to_i : nil
```
Comment (liam-mitchell) [inline messages.rb]: Generally the style in inne is to leave spaces around binary operators like this, so `20 - rank.to_i`

File: messages.rb [hunk 3]
```diff
@@ -260,14 +265,22 @@ def send_list(event)
   player = parse_player(msg, event.user.name)
   type = parse_type(msg)
   tabs = parse_tabs(msg)
+  rank = parse_rank(msg) || 20
+  bott = parse_bottom_rank(msg) || 0
+  ties = !!(msg =~ /ties/i)
   all = player.scores_by_rank(type, tabs)
 
+  if rank==20 && bott==0 && !!msg[/0th/i]
```
Comment (liam-mitchell) [inline messages.rb]: spaces again :)

File: messages.rb [hunk 4]
```diff
@@ -276,6 +289,92 @@ def send_list(event)
   event.attach_file(File::open(tmpfile))
 end
 
+def send_community(event)
+  msg = event.content
+  tabs = parse_tabs(msg)
+  tabs.empty? ? ftabs = "" : ftabs = format_tabs(tabs)
```
Comment (liam-mitchell) [inline messages.rb]: I think `format_tabs` will take care of leaving this string empty if the array is empty. So no need to check here.

Also, if you're going to assign to a variable in a ternary like this, it's more pleasant to put it after the `=` like `ftabs = tabs.empty? "" : format_tabs(tabs)`. You avoid repeating the variable name, and make it clear that it's an assignment - I didn't realize this declared a new variable for a minute.

File: messages.rb [hunk 5]
```diff
@@ -276,6 +289,92 @@ def send_list(event)
...
+  tabs = (tabs.empty? ? [:SI, :S, :SL, :SU, :SS, :SS2] : tabs)
+
+  levels = Score.where(rank: 0, highscoreable_type: "Level").includes(:level).where(levels: {tab: tabs})
+  episodes = Score.where(rank: 0, highscoreable_type: "Episode").includes(:episode).where(episodes: {tab: tabs})
+  lvls = levels.count
```
Comment (liam-mitchell) [inline messages.rb]: I find these names a bit confusing since `levels` and `lvls` are basically the same term, same with `episodes` and `eps`. Maybe `n_levels` instead?

File: messages.rb [hunk 6]
```diff
@@ -276,6 +289,92 @@ def send_list(event)
...
+  eps = episodes.count
+  tls = levels.map{|s| s.score}.sum
+  tes = episodes.map{|s| s.score}.sum
+  if !(tabs&[:SS, :SS2]).empty? # in order to compare lvls and eps we have to exclude secrets
```
Comment (liam-mitchell) [inline messages.rb]: Cute. I didn't know you could use `&` on arrays like this, I like it.

File: messages.rb [hunk 7]
```diff
@@ -276,6 +289,92 @@ def send_list(event)
...
+  msg = event.content
+  tabs = parse_tabs(msg)
+  tabs.empty? ? ftabs = "" : ftabs = format_tabs(tabs)
+  tabs = (tabs.empty? ? [:SI, :S, :SL, :SU, :SS, :SS2] : tabs)
+
+  levels = Score.where(rank: 0, highscoreable_type: "Level").includes(:level).where(levels: {tab: tabs})
+  episodes = Score.where(rank: 0, highscoreable_type: "Episode").includes(:episode).where(episodes: {tab: tabs})
+  lvls = levels.count
+  eps = episodes.count
+  tls = levels.map{|s| s.score}.sum
```
Comment (liam-mitchell) [inline messages.rb]: 1. I prefer spaces around blocks like this as well, `map { |s|` instead of `map{|s|`.
2. `tls` and `tes` is a bit challenging to understand - total level/episode score, presumably? Maybe just `level_total`/`episode_total` instead?

File: messages.rb [hunk 8]
```diff
@@ -276,6 +289,92 @@ def send_list(event)
...
+  tes = episodes.map{|s| s.score}.sum
+  if !(tabs&[:SS, :SS2]).empty? # in order to compare lvls and eps we have to exclude secrets
+    tlsno = Score.where(rank: 0, highscoreable_type: "Level").includes(:level).where(levels: {tab: tabs}).where.not(levels: {tab: [:SS, :SS2]}).map{|s| s.score}.sum
+    ttlsno = "Total level score (w/o secrets): #{tlsno}\n"
+  else
+    tlsno = tls
+    ttlsno = ""
+  end
+  diff = tlsno-4*90*eps-tes
```
Comment (liam-mitchell) [inline messages.rb]: I'm a bit confused about what this difference is computing exactly. What's the goal here?

File: messages.rb [hunk 9]
```diff
@@ -276,6 +289,92 @@ def send_list(event)
...
+  ttls = "Total level score: #{tls}\n"
+  ttes = "Total episode score: #{tes}\n"
+  tdiff = "Difference: #{"%.3f" % [diff]}\n"
+  ttlsavg = "Average level score: #{"%.3f" % [tls/lvls]}\n"
+  ttesavg = "Average episode score: #{"%.3f" % [tes/eps]}\n"
+  avgdiff = "Average difference: #{"%.3f" % [diff/eps]}\n"
+
+  event << "Community's total #{ftabs}scores #{Time.now.strftime("on %A %B %-d at %H:%M:%S (%z)")}:\n```#{ttls}#{ttlsno}#{ttes}#{tdiff}\n#{ttlsavg}#{ttesavg}#{avgdiff}```"
```
Comment (liam-mitchell) [inline messages.rb]: This `strftime` with that format should probably be broken out into its own `format_time` or something like that since it's repeated a few times now.

File: messages.rb [hunk 10]
```diff
@@ -276,6 +289,92 @@ def send_list(event)
...
+    tlsno = tls
+    ttlsno = ""
+  end
+  diff = tlsno-4*90*eps-tes
+  ttls = "Total level score: #{tls}\n"
+  ttes = "Total episode score: #{tes}\n"
+  tdiff = "Difference: #{"%.3f" % [diff]}\n"
+  ttlsavg = "Average level score: #{"%.3f" % [tls/lvls]}\n"
+  ttesavg = "Average episode score: #{"%.3f" % [tes/eps]}\n"
+  avgdiff = "Average difference: #{"%.3f" % [diff/eps]}\n"
```
Comment (liam-mitchell) [inline messages.rb]: Most of these variables suffer a bit from the same problem as above - the abbreviations are a bit arcane.

File: messages.rb [hunk 11]
```diff
@@ -547,6 +646,10 @@ def respond(event)
   send_level_name(event) if msg =~ /\blevel name\b/i
   send_level_id(event) if msg =~ /\blevel id\b/i
   send_diff(event) if msg =~ /diff/i
+  send_community(event) if msg =~ /community/i && msg !~ /retirement/i
```
Comment (liam-mitchell) [inline messages.rb]: `"retirement"`?

File: models.rb [hunk 12]
```diff
@@ -91,6 +107,10 @@ def spread(n)
     scores.find_by(rank: n).spread unless !scores.exists?(rank: n)
   end
 
+  def tie
```
Comment (liam-mitchell) [inline models.rb]: I'd prefer the name `tie_count` or similar, since this doesn't return 'a tie' but instead the number of ties for the highscore.

File: models.rb [hunk 13]
```diff
@@ -122,6 +142,10 @@ class Level < ActiveRecord::Base
   has_many :scores, as: :highscoreable
   enum tab: [:SI, :S, :SU, :SL, :SS, :SS2]
 
+  def self.count
```
Comment (liam-mitchell) [inline models.rb]: I believe ActiveRecord has something like `Level.count` built in, so you shouldn't need to do this spooky `ObjectSpace` counting stuff :)

**Comment by liam-mitchell** [comment]
One general comment: can you add README entries for these commands? They'll show up in `help` as well if you format them correctly :)

File: messages.rb (hunk 2 above)
Comment (edelkas) [reply messages.rb]: Fixed! I'll keep it in mind from now on.

File: messages.rb (hunk 3 above)
Comment (edelkas) [reply messages.rb]: Fixed ;)

File: messages.rb (hunk 4 above)
Comment (edelkas) [reply messages.rb]: Oh, you're right. It was giving me problems because, if tabs is empty, I fill it afterwards in order to always be able to use the line "where(tab: tabs)".

I've changed this method to what seems to be more customary in inne, which is to define a static method in the module (Score, in this case) that takes care of this, and use it to return the tls and the tes. I think the code looks cleaner now, since this method looked a bit convoluted.

File: messages.rb (hunk 6 above)
Comment (edelkas) [reply messages.rb]: Learned it just for this, apart from the intersection, one can get the union with `|` and the difference with `-`.

File: messages.rb (hunk 5 above)
Comment (edelkas) [reply messages.rb]: True, I've changed all variable names to make them clearer, I've also significantly reduced the number of variables by using a single string and appending to it. Also, since I've revamped this method and it now depends on a static Score class method, there's even less trouble with variables now, so the method is much cleaner.

File: messages.rb (hunk 7 above)
Comment (edelkas) [reply messages.rb]: Gotcha, I'll keep that spacing in mind ;)
Also updated all those weird variable names.

File: messages.rb (hunk 8 above)
Comment (edelkas) [reply messages.rb]: Oh yeah, the goal is to see how much time is being lost in the episode 0ths with respect to the level 0ths, both in total, and later on also in average. In other words, how clean episodes are globally. This function was also present in NHigh (albeit less complete).

However, we have to take into account that each level gives you 90 seconds at the start, but when playing an episode, only the first level actually does this, and you have to play the remaining levels with the time you've accumulated up to that point, as of course, you already know.

This means that the episode score will always be 4 * 90 seconds lower than the sum of all 5 level scores, so if you want to compare both (e.g. to see how clean the episode is), you will have to subtract that amount.

Naturally this has the same effect when we find the total sum, so we will have to subtract that amount times the number of episodes there are. This is how we find the true difference (or the true amount of time being lost in episodes), which we can later average as well :)

File: messages.rb (hunk 10 above)
Comment (edelkas) [reply messages.rb]: Indeed, clarified all of that mess as already explained on a previous comment.

File: messages.rb (hunk 11 above)
Comment (edelkas) [reply messages.rb]: Oh hah, yeah, I expected that to be enigmatic. I tried to define keywords that wouldn't accidentally call other methods as well. There's a single level with the word "community" in its name, called "cyberdemon retirement community". So this way, I prevent this method from being called when people ask inne about stuff regarding that level.

This is also the problem that is faced, for instance, when people ask for the levels "a frankly unbelievable amount of doors" or "s-rank" and the 0th rankings are also called, as noted by issue #41, except this one can't be fixed even with word boundaries. Another example is that there are 4 levels with the word "history" in their name.

As a consequence, I thought about making inne's commands bullet proof. Adding the word boundaries would be the first step, as noted by issue #41, and I might get on that sometime. But that still leaves problematic cases like this one, and since there's few of those, I thought a more definitive solution would be to address them individually like here.

I admit it doesn't look very elegant but I can't think of anything else, however, if you deem this unnecessary or a rather crude solution, I can ditch it no problem ;)

PS. A possible alteration of this when there's many level names to exclude could be to, instead, exclude the problematic words "screenshot", "scores", etc. Although excluding "scores" is usually not a good idea, like in this case, where it's likely people will write "community total scores".

File: models.rb (hunk 12 above)
Comment (edelkas) [reply models.rb]: True, fixed.

File: models.rb (hunk 13 above)
Comment (edelkas) [reply models.rb]: Whoops you're right, I don't know why I didn't try / search that at first.

File: messages.rb (hunk 9 above)
Comment (edelkas) [reply messages.rb]: Done, that really cleans up the code.

**Comment by edelkas** [comment]
The message length of inne's help was already almost maxed, so after this addition (and the potential future ones), I've had to cut the help into two messages.

The most elegant way I thought for doing this was to add an imperceptible space in the (already blank) line from the readme you want inne to break the message. This way you could potentially add as many breaks simply into the readme.

I thought this was better than some other automatic possibilities based on the actual char count limit, that would probably break the help message on weird spots. This way I pretty much halved it.

File: messages.rb (hunk 11 above)
Comment (liam-mitchell) [reply messages.rb]: Ahh. That makes sense.

One possibility is maybe to trim level names and usernames (so things after the word 'for', mostly) before doing this match. I think those are likely to be the only things that cause problems, though  I may be wrong about that... either way, it would avoid all of those problems, which are always embarrassing when they come up ;)

Either way, if we do leave it like this, 'I expected that to be enigmatic' means it's a prime spot for a nice comment in my opinion.

File: messages.rb (hunk 6 above)
Comment (liam-mitchell) [reply messages.rb]: I've used `|` before, but never extended the concept in my brain to the logical conclusion. Nice.

File: messages.rb (hunk 8 above)
Comment (liam-mitchell) [reply messages.rb]: Ah, that makes sense, cool. Maybe a quick comment here explaining, and you may want to add a bit of explanation to the output as well (instead of just 'Average difference:' which is a bit opaque) - though I'm having a hard time thinking of how you would describe this without being way too wordy...

'Average difference between level and episode 0ths:' maybe? Even that's pretty long, but it at least attempts to explain what's being computed.

**Comment by liam-mitchell** [comment]
Great, nice that you figured out something that works nicely for arbitrary length README :) and is easy to format nicely for us as well.

File: messages.rb (hunk 8 above)
Comment (edelkas) [reply messages.rb]: Yeah true, I was too used to the term, clarified the code and output.

**Comment by liam-mitchell** [comment]
Thanks so much! Merged it in, it's live in Discord now :)
